svm/src/account_loader.rs

```diff
@@ -2227,4 +2371,261 @@ mod tests {
+
+            // fee-payer counted once
+            let ixns = vec![Instruction::new_with_bytes(
+                program1,
+                &[],
+                vec![account_meta(fee_payer, false)],
+            )];
+            test_data_size(ixns, program1_size + bpf_loader_size + fee_payer_size);
+
+            // edge cases involving program cache
+            let mut program_cache = ProgramCacheForTxBatch::default();
+
+            let program2_entry = ProgramCacheEntry {
+                account_size: (program2_size + programdata2_size) as usize,
+                account_owner: ProgramCacheEntryOwner::LoaderV3,
+                ..ProgramCacheEntry::default()
+            };
+            program_cache.replenish(program2, Arc::new(program2_entry));
+
+            // normal function call uses the combined cache size
+            let ixns = vec![Instruction::new_with_bytes(program2, &[], vec![])];
+            test_data_size_with_cache(
+                ixns,
+                program_cache.clone(),
+                program2_size + programdata2_size + upgradeable_loader_size + fee_payer_size,
+            );
+
+            // program as instruction account bypasses the cache
+            let ixns = vec![Instruction::new_with_bytes(
+                program2,
+                &[],
+                vec![account_meta(program2, false)],
+            )];
+            test_data_size_with_cache(
+                ixns,
+                program_cache.clone(),
+                program2_size + upgradeable_loader_size + fee_payer_size,
+            );
+
+            // programdata as instruction account double-counts it
+            let ixns = vec![Instruction::new_with_bytes(
+                program2,
+                &[],
+                vec![account_meta(programdata2, false)],
+            )];
+            test_data_size_with_cache(
+                ixns,
+                program_cache.clone(),
+                program2_size + programdata2_size * 2 + upgradeable_loader_size + fee_payer_size,
+            );
+
+            // both as instruction accounts, for completeness
+            let ixns = vec![Instruction::new_with_bytes(
+                program2,
+                &[],
+                vec![
+                    account_meta(program2, false),
+                    account_meta(programdata2, false),
+                ],
+            )];
+            test_data_size_with_cache(
+                ixns,
+                program_cache.clone(),
+                program2_size + programdata2_size + upgradeable_loader_size + fee_payer_size,
+            );
+
+            // NOTE for the new loader we *must* also test arbitrary permutations of the cache transactions
+            // to ensure that the batched loading is overridden on a tx-per-tx basis
+            // also we will need to create a fake vm and use ProgramCacheEntryType::Loaded
+            // since the old loader does not check tombstones, but the new one does
```

Comment:
i deleted this part of the comment since it is no longer relevant (i previously wrote a batched account loader that checked tombstones with an escape hatch to emulate old behavior, i now have a jit account loader which retains the old behavior unmodified), but sorry this was unclear, i meant the old and new account loaders, not the bpf loaders